src/btree/bt_discard.c

```diff
@@ -245,20 +245,31 @@ __free_page_modify(WT_SESSION_IMPL *session, WT_PAGE *page)
 void
 __wt_ref_addr_free(WT_SESSION_IMPL *session, WT_REF *ref)
 {
+    WT_PAGE *home;
     void *ref_addr;
 
     /*
-     * The page being discarded may be the child of a page being split, where the WT_REF.addr field
-     * is being instantiated (as it can no longer reference the on-disk image). Loop until we read
-     * and clear the address without a race, then free the read address as necessary.
+     * In order to free the WT_REF.addr field we need to read and clear the address without a race.
+     * The WT_REF may be a child of a page being split, in which case the addr field could be
+     * instantiated concurrently which changes the addr field. Once we swap in NULL we effectively
+     * own the addr. Then provided the addr is off page we can free the memory.
+     *
+     * However as we could be the child of a page being split the ref->home pointer which tells us
+     * whether the addr is on or off page could change concurrently. To avoid this we save the home
+     * pointer before we do the compare and swap. Additionally ref->home is marked as volatile so
```

Comment:
We don't use volatile in WiredTiger for that purpose at the moment. If it's something this code relies on, then using a WT_ORDERED_READ would be the right choice. I suspect the subsequent ordered read into ref_addr does the job, but we might as well be exceedingly cautious.

Reply:
> We don't use volatile in WiredTiger for that purpose at the moment

That's fair, perhaps `WT_BARRIER` is a better fit since we only need a compiler barrier here. 

> I suspect the subsequent ordered read into ref_addr does the job, but we might as well be exceedingly cautious.

My understanding that it definitely does the job - we use a full memory fence, so the `ref->home` read can't be ordered past the `ref->addr` read.

Reply:
> My understanding that it definitely does the job - we use a memory fence rather than one of the instructions that's associated with a single read, so the ref->home read can't be ordered past the ref->addr read.

I agree with @korteland  on that one. It sounds like a `WT_BARRIER` is sufficient.

Reply:
[52817b9](https://github.com/wiredtiger/wiredtiger/pull/8870/commits/52817b9aeb4a08e15433f58bf2f17c5012626eb8)